In [ ]:
from google.colab import drive
drive.mount('drive')

Drive already mounted at drive; to attempt to forcibly remount, call drive.mount("drive", force_remount=True).


In [ ]:
import pandas as pd

In [ ]:
dataset = pd.read_csv('drive/My Drive/Internshala Assignment/Accredian/Fraud.csv')
dataset

step      type      amount     nameOrig  oldbalanceOrg  \
0           1   PAYMENT     9839.64  C1231006815      170136.00   
1           1   PAYMENT     1864.28  C1666544295       21249.00   
2           1  TRANSFER      181.00  C1305486145         181.00   
3           1  CASH_OUT      181.00   C840083671         181.00   
4           1   PAYMENT    11668.14  C2048537720       41554.00   
...       ...       ...         ...          ...            ...   
6362615   743  CASH_OUT   339682.13   C786484425      339682.13   
6362616   743  TRANSFER  6311409.28  C1529008245     6311409.28   
6362617   743  CASH_OUT  6311409.28  C1162922333     6311409.28   
6362618   743  TRANSFER   850002.52  C1685995037      850002.52   
6362619   743  CASH_OUT   850002.52  C1280323807      850002.52   

         newbalanceOrig     nameDest  oldbalanceDest  newbalanceDest  isFraud  \
0             160296.36  M1979787155            0.00            0.00        0   
1              19384.72  M2044282225            0.00            0.00        0   
2                  0.00   C553264065            0.00            0.00        1   
3                  0.00    C38997010        21182.00            0.00        1   
4              29885.86  M1230701703            0.00            0.00        0   
...                 ...          ...             ...             ...      ...   
6362615            0.00   C776919290            0.00       339682.13        1   
6362616            0.00  C1881841831            0.00            0.00        1   
6362617            0.00  C1365125890        68488.84      6379898.11        1   
6362618            0.00  C2080388513            0.00            0.00        1   
6362619            0.00   C873221189      6510099.11      7360101.63        1   

         isFlaggedFraud  
0                     0  
1                     0  
2                     0  
3                     0  
4                     0  
...                 ...  
6362615               0  
6362616               0  
6362617               0  
6362618               0  
6362619               0  

[6362620 rows x 11 columns]

In [ ]:
dataset=dataset.dropna()

In [ ]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6362620 entries, 0 to 6362619
Data columns (total 11 columns):
 #   Column          Dtype  
---  ------          -----  
 0   step            int64  
 1   type            object 
 2   amount          float64
 3   nameOrig        object 
 4   oldbalanceOrg   float64
 5   newbalanceOrig  float64
 6   nameDest        object 
 7   oldbalanceDest  float64
 8   newbalanceDest  float64
 9   isFraud         int64  
 10  isFlaggedFraud  int64  
dtypes: float64(5), int64(3), object(3)
memory usage: 534.0+ MB


In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
X = dataset.iloc[:,:9]
y = dataset.iloc[:,9]

In [ ]:
from sklearn.preprocessing import LabelEncoder,OneHotEncoder

labelencoder_X = LabelEncoder()
X['type'] = labelencoder_X.fit_transform(X['type'])
X

step  type      amount     nameOrig  oldbalanceOrg  newbalanceOrig  \
0           1     3     9839.64  C1231006815      170136.00       160296.36   
1           1     3     1864.28  C1666544295       21249.00        19384.72   
2           1     4      181.00  C1305486145         181.00            0.00   
3           1     1      181.00   C840083671         181.00            0.00   
4           1     3    11668.14  C2048537720       41554.00        29885.86   
...       ...   ...         ...          ...            ...             ...   
6362615   743     1   339682.13   C786484425      339682.13            0.00   
6362616   743     4  6311409.28  C1529008245     6311409.28            0.00   
6362617   743     1  6311409.28  C1162922333     6311409.28            0.00   
6362618   743     4   850002.52  C1685995037      850002.52            0.00   
6362619   743     1   850002.52  C1280323807      850002.52            0.00   

            nameDest  oldbalanceDest  newbalanceDest  
0        M1979787155            0.00            0.00  
1        M2044282225            0.00            0.00  
2         C553264065            0.00            0.00  
3          C38997010        21182.00            0.00  
4        M1230701703            0.00            0.00  
...              ...             ...             ...  
6362615   C776919290            0.00       339682.13  
6362616  C1881841831            0.00            0.00  
6362617  C1365125890        68488.84      6379898.11  
6362618  C2080388513            0.00            0.00  
6362619   C873221189      6510099.11      7360101.63  

[6362620 rows x 9 columns]

In [ ]:
X['totalTransaction']=X['oldbalanceOrg']-X['newbalanceOrig']
X

step  type      amount     nameOrig  oldbalanceOrg  newbalanceOrig  \
0           1     3     9839.64  C1231006815      170136.00       160296.36   
1           1     3     1864.28  C1666544295       21249.00        19384.72   
2           1     4      181.00  C1305486145         181.00            0.00   
3           1     1      181.00   C840083671         181.00            0.00   
4           1     3    11668.14  C2048537720       41554.00        29885.86   
...       ...   ...         ...          ...            ...             ...   
6362615   743     1   339682.13   C786484425      339682.13            0.00   
6362616   743     4  6311409.28  C1529008245     6311409.28            0.00   
6362617   743     1  6311409.28  C1162922333     6311409.28            0.00   
6362618   743     4   850002.52  C1685995037      850002.52            0.00   
6362619   743     1   850002.52  C1280323807      850002.52            0.00   

        nameDest  oldbalanceDest  newbalanceDest  totalRemoved  \
0              M            0.00            0.00       9839.64   
1              M            0.00            0.00       1864.28   
2              C            0.00            0.00        181.00   
3              C        21182.00            0.00        181.00   
4              M            0.00            0.00      11668.14   
...          ...             ...             ...           ...   
6362615        C            0.00       339682.13     339682.13   
6362616        C            0.00            0.00    6311409.28   
6362617        C        68488.84      6379898.11    6311409.28   
6362618        C            0.00            0.00     850002.52   
6362619        C      6510099.11      7360101.63     850002.52   

         totalTransaction  
0                 9839.64  
1                 1864.28  
2                  181.00  
3                  181.00  
4                11668.14  
...                   ...  
6362615         339682.13  
6362616        6311409.28  
6362617        6311409.28  
6362618         850002.52  
6362619         850002.52  

[6362620 rows x 11 columns]

In [ ]:
def initial(nameDest):
  return nameDest[0]


In [ ]:
X['nameDest'] = X['nameDest'].apply(initial)
X

step  type      amount     nameOrig  oldbalanceOrg  newbalanceOrig  \
0           1     3     9839.64  C1231006815      170136.00       160296.36   
1           1     3     1864.28  C1666544295       21249.00        19384.72   
2           1     4      181.00  C1305486145         181.00            0.00   
3           1     1      181.00   C840083671         181.00            0.00   
4           1     3    11668.14  C2048537720       41554.00        29885.86   
...       ...   ...         ...          ...            ...             ...   
6362615   743     1   339682.13   C786484425      339682.13            0.00   
6362616   743     4  6311409.28  C1529008245     6311409.28            0.00   
6362617   743     1  6311409.28  C1162922333     6311409.28            0.00   
6362618   743     4   850002.52  C1685995037      850002.52            0.00   
6362619   743     1   850002.52  C1280323807      850002.52            0.00   

        nameDest  oldbalanceDest  newbalanceDest  totalRemoved  \
0              M            0.00            0.00       9839.64   
1              M            0.00            0.00       1864.28   
2              C            0.00            0.00        181.00   
3              C        21182.00            0.00        181.00   
4              M            0.00            0.00      11668.14   
...          ...             ...             ...           ...   
6362615        C            0.00       339682.13     339682.13   
6362616        C            0.00            0.00    6311409.28   
6362617        C        68488.84      6379898.11    6311409.28   
6362618        C            0.00            0.00     850002.52   
6362619        C      6510099.11      7360101.63     850002.52   

         totalTransaction  
0                 9839.64  
1                 1864.28  
2                  181.00  
3                  181.00  
4                11668.14  
...                   ...  
6362615         339682.13  
6362616        6311409.28  
6362617        6311409.28  
6362618         850002.52  
6362619         850002.52  

[6362620 rows x 11 columns]

In [ ]:
X['nameOrig']=X['nameOrig'].apply(initial)
X

step  type      amount nameOrig  oldbalanceOrg  newbalanceOrig  \
0           1     3     9839.64        C      170136.00       160296.36   
1           1     3     1864.28        C       21249.00        19384.72   
2           1     4      181.00        C         181.00            0.00   
3           1     1      181.00        C         181.00            0.00   
4           1     3    11668.14        C       41554.00        29885.86   
...       ...   ...         ...      ...            ...             ...   
6362615   743     1   339682.13        C      339682.13            0.00   
6362616   743     4  6311409.28        C     6311409.28            0.00   
6362617   743     1  6311409.28        C     6311409.28            0.00   
6362618   743     4   850002.52        C      850002.52            0.00   
6362619   743     1   850002.52        C      850002.52            0.00   

        nameDest  oldbalanceDest  newbalanceDest  totalRemoved  \
0              M            0.00            0.00       9839.64   
1              M            0.00            0.00       1864.28   
2              C            0.00            0.00        181.00   
3              C        21182.00            0.00        181.00   
4              M            0.00            0.00      11668.14   
...          ...             ...             ...           ...   
6362615        C            0.00       339682.13     339682.13   
6362616        C            0.00            0.00    6311409.28   
6362617        C        68488.84      6379898.11    6311409.28   
6362618        C            0.00            0.00     850002.52   
6362619        C      6510099.11      7360101.63     850002.52   

         totalTransaction  
0                 9839.64  
1                 1864.28  
2                  181.00  
3                  181.00  
4                11668.14  
...                   ...  
6362615         339682.13  
6362616        6311409.28  
6362617        6311409.28  
6362618         850002.52  
6362619         850002.52  

[6362620 rows x 11 columns]

In [ ]:
from sklearn.preprocessing import LabelEncoder,OneHotEncoder

labelencoder_X = LabelEncoder()
X['nameOrig'] = labelencoder_X.fit_transform(X['nameOrig'])
X

step  type      amount  nameOrig  oldbalanceOrg  newbalanceOrig  \
0           1     3     9839.64         0      170136.00       160296.36   
1           1     3     1864.28         0       21249.00        19384.72   
2           1     4      181.00         0         181.00            0.00   
3           1     1      181.00         0         181.00            0.00   
4           1     3    11668.14         0       41554.00        29885.86   
...       ...   ...         ...       ...            ...             ...   
6362615   743     1   339682.13         0      339682.13            0.00   
6362616   743     4  6311409.28         0     6311409.28            0.00   
6362617   743     1  6311409.28         0     6311409.28            0.00   
6362618   743     4   850002.52         0      850002.52            0.00   
6362619   743     1   850002.52         0      850002.52            0.00   

        nameDest  oldbalanceDest  newbalanceDest  totalRemoved  \
0              M            0.00            0.00       9839.64   
1              M            0.00            0.00       1864.28   
2              C            0.00            0.00        181.00   
3              C        21182.00            0.00        181.00   
4              M            0.00            0.00      11668.14   
...          ...             ...             ...           ...   
6362615        C            0.00       339682.13     339682.13   
6362616        C            0.00            0.00    6311409.28   
6362617        C        68488.84      6379898.11    6311409.28   
6362618        C            0.00            0.00     850002.52   
6362619        C      6510099.11      7360101.63     850002.52   

         totalTransaction  
0                 9839.64  
1                 1864.28  
2                  181.00  
3                  181.00  
4                11668.14  
...                   ...  
6362615         339682.13  
6362616        6311409.28  
6362617        6311409.28  
6362618         850002.52  
6362619         850002.52  

[6362620 rows x 11 columns]

In [ ]:
from sklearn.preprocessing import LabelEncoder,OneHotEncoder

labelencoder_X = LabelEncoder()
X['nameDest'] = labelencoder_X.fit_transform(X['nameDest'])
X

step  type      amount  nameOrig  oldbalanceOrg  newbalanceOrig  \
0           1     3     9839.64         0      170136.00       160296.36   
1           1     3     1864.28         0       21249.00        19384.72   
2           1     4      181.00         0         181.00            0.00   
3           1     1      181.00         0         181.00            0.00   
4           1     3    11668.14         0       41554.00        29885.86   
...       ...   ...         ...       ...            ...             ...   
6362615   743     1   339682.13         0      339682.13            0.00   
6362616   743     4  6311409.28         0     6311409.28            0.00   
6362617   743     1  6311409.28         0     6311409.28            0.00   
6362618   743     4   850002.52         0      850002.52            0.00   
6362619   743     1   850002.52         0      850002.52            0.00   

         nameDest  oldbalanceDest  newbalanceDest  totalRemoved  \
0               1            0.00            0.00       9839.64   
1               1            0.00            0.00       1864.28   
2               0            0.00            0.00        181.00   
3               0        21182.00            0.00        181.00   
4               1            0.00            0.00      11668.14   
...           ...             ...             ...           ...   
6362615         0            0.00       339682.13     339682.13   
6362616         0            0.00            0.00    6311409.28   
6362617         0        68488.84      6379898.11    6311409.28   
6362618         0            0.00            0.00     850002.52   
6362619         0      6510099.11      7360101.63     850002.52   

         totalTransaction  
0                 9839.64  
1                 1864.28  
2                  181.00  
3                  181.00  
4                11668.14  
...                   ...  
6362615         339682.13  
6362616        6311409.28  
6362617        6311409.28  
6362618         850002.52  
6362619         850002.52  

[6362620 rows x 11 columns]

In [ ]:
from sklearn.model_selection import train_test_split
X_train , X_test , y_train , y_test = train_test_split(X,y,test_size=0.2,random_state=0)

In [ ]:
from xgboost import XGBClassifier
classifier = XGBClassifier()
classifier.fit(X_train , y_train)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, gpu_id=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=None, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              n_estimators=100, n_jobs=None, num_parallel_tree=None,
              predictor=None, random_state=None, ...)

In [ ]:
from sklearn.metrics import confusion_matrix, accuracy_score
y_pred = classifier.predict(X_test)
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[1270845      38]
 [    204    1437]]


0.9998098267694755

**Fraud Detection Model Description:**

The fraud detection model presented in this analysis is designed to identify fraudulent financial transactions.This model leverages various data preprocessing techniques and the XGBoost classification algorithm for fraud detection. Here's a comprehensive overview of the model:

**1. Data Loading and Cleaning:**
   - The analysis starts by loading the dataset from a CSV file stored in Google Drive.
   - Data cleaning is performed to ensure data quality, with missing values being removed from the dataset.



**2. Feature Engineering:**
   - Feature engineering is a crucial step in enhancing the model's predictive capabilities.
   - The 'type' column, representing transaction types, is encoded into numerical values using LabelEncoder.
   - The 'totalTransaction' column is created by calculating the difference between 'oldbalanceOrg' and 'newbalanceOrig'. This new feature could capture essential transaction behavior information.
   - The 'nameDest' and 'nameOrig' columns undergo preprocessing. A custom function, 'initial', is applied to these columns, likely extracting the first character to simplify categorical features.

**3. Label Encoding:**
   - Label encoding is applied to the 'nameOrig' and 'nameDest' columns using LabelEncoder to convert categorical data into a numeric format, which is essential for most machine learning algorithms.

**4. Data Splitting:**
   - The dataset is divided into training and testing sets using the `train_test_split` function. This separation ensures the model's ability to generalize to unseen data.

**5. Model Training:**
   - The model employs the XGBoost algorithm, known for its robust performance in classification tasks.
   - An XGBClassifier is initialized and trained on the training data. During training, the model learns to identify patterns and relationships in the features that aid in making predictions about the fraudulent nature of transactions.

**6. Model Evaluation:**
   - To assess the model's performance, predictions are made on the test data using the trained XGBoost model.
   - A confusion matrix is computed, breaking down the model's predictions into true positives, true negatives, false positives, and false negatives.
   - The primary evaluation metric used is the accuracy score, which measures the proportion of correct predictions made by the model on the test data. It provides an overall assessment of the model's effectiveness in identifying fraudulent transactions.

In summary, this fraud detection model is a supervised machine learning system designed to predict the likelihood of fraud in financial transactions. It leverages data preprocessing, label encoding, and the XGBoost algorithm to make predictions. The model's performance is evaluated using accuracy, providing an overall measure of its effectiveness in distinguishing fraudulent transactions from legitimate ones.

3) How did you select variables to be included in the model?

Variables in the model were likely selected based on domain knowledge, exploratory data analysis, and feature importance analysis. Key factors such as transaction amount, type, balances, and new features generated through feature engineering were considered. It's an iterative process guided by a balance between capturing important information and avoiding overfitting.

4)Demonstrate the performance of the model by using best set of tools.


 model is performing exceptionally well in terms of accuracy, indicating that it's making very accurate predictions.
 It correctly identifies the vast majority of non-fraudulent transactions (high TN count) and fraudulent transactions (high TP count).
 The small counts of false positives (38) and false negatives (204) indicate that model makes very few mistakes in classifying transactions.

5)What are the key factors that predict fraudulent customer?
  Do these factors make sense? If yes, How? If not, How not?


NewBalanceorig is the key factor . As it can be seen that in case of frauds , all amount from the account is Transacted to another account.

7)What kind of prevention should be adopted while company update its infrastructure?


 Ensure that the data format and features used for training fraud detection model remain consistent during the infrastructure update. Changes in data structure can affect the model's performance.
 After updating the infrastructure, it's crucial to retrain model using the most up-to-date data. Infrastructure updates may impact the model's behavior, so retraining helps maintain its accuracy.
 Before deploying the updated model, conduct rigorous testing and validation to confirm that it performs as expected. Verify that the accuracy and predictive power of the model remain consistent.

8) Assuming these actions have been implemented, how would you determine if they work?


Compare the model's performance metrics before and after the infrastructure update.
Perform cross-validation on your updated model with the new data to ensure it generalizes well and exhibits stable performance across different subsets of the data.
Use a separate hold-out test dataset that wasn't used during model training or validation to assess the model's real-world performance in the updated infrastructure.
